Update configuration management and enhance file structure, add test-matrix

AnkushMalaker opened the pull request:
- Refactored configuration file paths to use a dedicated `config/` directory, including updates to `config.yml` and its template.
- Modified service scripts to load the new configuration path for `config.yml`.
- Enhanced `.gitignore` to include the new configuration files and templates.
- Updated documentation to reflect changes in configuration file locations and usage.
- Improved setup scripts to ensure proper creation and management of configuration files.
- Added new test configurations for various provider combinations to streamline testing processes.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added multiple preconfigured test setups for common STT/LLM stacks.
  * Switched integration testing to Robot Framework with end-to-end memory validation and quality checks.

* **Chores**
  * Moved configuration into a dedicated config/ directory and updated templates.
  * Updated documentation and test scripts to use the new config paths and improve test isolation (unique project names, configurable test configs).

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

AnkushMalaker committed:
Add ConfigManager for unified configuration management

- Introduced a new `config_manager.py` module to handle reading and writing configurations from `config.yml` and `.env` files, ensuring backward compatibility.
- Refactored `ChronicleSetup` in `backends/advanced/init.py` to utilize `ConfigManager` for loading and updating configurations, simplifying the setup process.
- Removed redundant methods for loading and saving `config.yml` directly in `ChronicleSetup`, as these are now managed by `ConfigManager`.
- Enhanced user feedback during configuration updates, including success messages for changes made to configuration files.

AnkushMalaker committed:
Update configuration management and enhance file structure

- Refactored configuration file paths to use a dedicated `config/` directory, including updates to `config.yml` and its template.
- Modified service scripts to load the new configuration path for `config.yml`.
- Enhanced `.gitignore` to include the new configuration files and templates.
- Updated documentation to reflect changes in configuration file locations and usage.
- Improved setup scripts to ensure proper creation and management of configuration files.
- Added new test configurations for various provider combinations to streamline testing processes.

AnkushMalaker committed:
Add test requirements and clean up imports in wizard.py

- Introduced a new `test-requirements.txt` file to manage testing dependencies.
- Removed redundant import of `shutil` in `wizard.py` to improve code clarity.

AnkushMalaker committed:
Refactor transcription provider configuration and enhance setup process

- Updated `.env.template` to clarify speech-to-text configuration and removed deprecated options for Mistral.
- Modified `docker-compose.yml` to streamline environment variable management by removing unused Mistral keys.
- Enhanced `ChronicleSetup` in `init.py` to provide clearer user feedback and updated the transcription provider selection process to rely on `config.yml`.
- Improved error handling in the websocket controller to determine the transcription provider from the model registry instead of environment variables.
- Updated health check routes to reflect the new method of retrieving the transcription provider from `config.yml`.
- Adjusted `config.yml.template` to include comments on transcription provider options for better user guidance.

AnkushMalaker committed:
Enhance ConfigManager with deep merge functionality

- Updated the `update_memory_config` method to perform a deep merge of updates into the memory configuration, ensuring nested dictionaries are merged correctly.
- Added a new `_deep_merge` method to handle recursive merging of dictionaries, improving configuration management capabilities.

AnkushMalaker committed:
Refactor run-test.sh and enhance memory extraction tests

- Removed deprecated environment variable handling for TRANSCRIPTION_PROVIDER in `run-test.sh`, streamlining the configuration process.
- Introduced a new `run-custom.sh` script for executing Robot tests with custom configurations, improving test flexibility.
- Enhanced memory extraction tests in `audio_keywords.robot` and `memory_keywords.robot` to include detailed assertions and result handling.
- Updated `queue_keywords.robot` to fail fast if a job is in a 'failed' state when expecting 'completed', improving error handling.
- Refactored `test_env.py` to load environment variables with correct precedence, ensuring better configuration management.

AnkushMalaker committed:
unify tests to robot test, add some more clean up

AnkushMalaker commented:
@coderabbitai review

AnkushMalaker committed:
Update health check configuration in docker-compose-test.yml (#241)

- Increased the number of retries from 5 to 10 for improved resilience during service readiness checks.
- Extended the start period from 30s to 60s to allow more time for services to initialize before health checks commence.

AnkushMalaker committed:
Add step to create test configuration file in robot-tests.yml

- Introduced a new step in the GitHub Actions workflow to copy the test configuration file from tests/configs/deepgram-openai.yml to a new config/config.yml.
- Added logging to confirm the creation of the test config file, improving visibility during the test setup process.

AnkushMalaker committed:
remove cache step since not required

0xrushi commented on backends/advanced/init.py:
```diff
@@ -28,16 +28,29 @@ def __init__(self, args=None):
+        if str(repo_root) not in sys.path:
+            sys.path.insert(0, str(repo_root))
+
+        from config_manager import ConfigManager
```
can we move this line 40-44 at the top ?

AnkushMalaker committed:
coderabbit comments

0xrushi commented on config_manager.py:
```diff
@@ -0,0 +1,364 @@
+        current = Path.cwd()
+
+        # Walk up until we find config.yml
+        while current != current.parent:
```
why is this while needed?

AnkushMalaker replied:
Its just so if this module is initialized without repo_root, it can walk up the tree to find it
I guess its an unbound upwards search, which has issues. 
Hmm... one could argue config_manager is always going to be at repo root so I can just do `__file__`?

0xrushi replied:
I'd stick with `__file__` unless there is some requirement for while.

0xrushi commented on config_manager.py:
```diff
@@ -0,0 +1,364 @@
+        """Load config.yml file."""
+        if not self.config_yml_path.exists():
+            logger.warning(f"config.yml not found at {self.config_yml_path}")
+            return {}
```
Could we raise a RuntimeError if file is not found?

0xrushi commented on config_manager.py:
```diff
@@ -0,0 +1,364 @@
+                return yaml.safe_load(f) or {}
+        except Exception as e:
+            logger.error(f"Failed to load config.yml: {e}")
+            return {}
```
Raise RuntimeError for invalid config

AnkushMalaker committed:
Refactor ConfigManager error handling for configuration file loading

- Updated the ConfigManager to raise RuntimeError exceptions when the configuration file is not found or is invalid, improving error visibility and user guidance.
- Removed fallback behavior that previously returned the current directory, ensuring users are explicitly informed about missing or invalid configuration files.

AnkushMalaker committed:
Refactor _find_repo_root method in ConfigManager

- Updated the _find_repo_root method to locate the repository root using the __file__ location instead of searching for config/config.yml, simplifying the logic and improving reliability.
- Removed the previous error handling that raised a RuntimeError if the configuration file was not found, as the new approach assumes config_manager.py is always at the repo root.

AnkushMalaker committed:
Merge branch 'fix/cleanup-pre-release-v0.1.0' into feat/test-matrix
